File: src/main/java/org/dxworks/githubminer/invitations/GithubInvitationService.java
package org.dxworks.githubminer.invitations;

import com.google.api.client.http.GenericUrl;
import com.google.api.client.http.HttpResponse;
import com.google.common.collect.ImmutableMap;
import com.google.common.reflect.TypeToken;
import lombok.SneakyThrows;
import org.dxworks.githubminer.GithubApiService;
import org.dxworks.githubminer.dto.response.invitations.Invitation;
import org.dxworks.utils.java.rest.client.providers.BasicAuthenticationProvider;

import java.lang.reflect.Type;
import java.math.BigDecimal;
import java.util.List;

public class GithubInvitationService extends GithubApiService {
    public GithubInvitationService() {
        super();
    }

    public GithubInvitationService(BasicAuthenticationProvider authenticationProvider) {
        super(authenticationProvider);
    }

    @SneakyThrows
    public List<Invitation> listInvitationsForUser() {
        String apiPath = getApiPath("user", "repository_invitations");

        HttpResponse httpResponse = httpClient.get(new GenericUrl(apiPath));

        Type type = new TypeToken<List<Invitation>>() {
        }.getType();
        return (List<Invitation>) httpResponse.parseAs(type);
    }

    public boolean acceptInvitation(Invitation invitation) {
        return acceptInvitation(invitation.getId());
    }

    public boolean acceptInvitation(BigDecimal invitationId) {
        ImmutableMap<String, String> map = ImmutableMap.of("invitation_id", invitationId.toString());
        String apiPath = getApiPath(map, "user", "repository_invitations", ":invitation_id");

        HttpResponse httpResponse = httpClient.patch(new GenericUrl(apiPath), null);

        return httpResponse.getStatusCode() == 204;
    }
}

File: src/main/kotlin/org/dxworks/githubminer/service/invitations/GithubInvitationService.kt
package org.dxworks.githubminer.service.invitations

import com.google.api.client.http.GenericUrl
import com.google.common.collect.ImmutableMap
import com.google.common.reflect.TypeToken
import org.dxworks.githubminer.constants.ANONYMOUS
import org.dxworks.githubminer.constants.GITHUB_API_PATH
import org.dxworks.githubminer.dto.response.invitations.Invitation
import org.dxworks.githubminer.service.GithubApiService
import java.math.BigDecimal

class GithubInvitationService(githubBasePath: String = GITHUB_API_PATH,
                              githubTokens: List<String> = listOf(ANONYMOUS))
    : GithubApiService(githubBasePath, githubTokens) {

    fun listInvitationsForUser(): List<Invitation> {
        val apiPath = getApiPath("user", "repository_invitations")
        val httpResponse = httpClient.get(GenericUrl(apiPath))
        val type = object : TypeToken<List<Invitation?>?>() {}.type
        return httpResponse.parseAs(type) as List<Invitation>
    }

    fun acceptInvitation(invitation: Invitation): Boolean {
        return acceptInvitation(invitation.id)
    }

    fun acceptInvitation(invitationId: BigDecimal): Boolean {
        val map = ImmutableMap.of("invitation_id", invitationId.toString())
        val apiPath = getApiPath(map, "user", "repository_invitations", ":invitation_id")
        val httpResponse = httpClient.patch(GenericUrl(apiPath), null)
        return httpResponse.statusCode == 204
    }
}
